src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -1502,3 +1530,217 @@ RelationUsesHeapAccessMethodOrNone(Relation relation)
+ * any references to or from the table.
+ *
+ * The dropping of old table is done with CASCADE. Anything not mentioned here will
+ * be dropped.
+ */
+void
+UndistributeTable(Oid relationId)
+{
+	Relation relation = try_relation_open(relationId, ExclusiveLock);
+	if (relation == NULL)
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute table"),
+						errdetail("No such distributed table exists. "
+								  "Might have already been undistributed.")));
+	}
+
+	relation_close(relation, NoLock);
+
+	if (!IsCitusTable(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute table."),
+						errdetail("The table is not distributed.")));
+	}
+
+	if (TableReferencing(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute table "
+							   "because it has a foreign key.")));
+	}
+
+	if (TableReferenced(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute table "
+							   "because a foreign key references to it.")));
+	}
+
+
+	List *tableCreationCommands = GetTableCreationCommands(relationId, true);
+	List *tableConstructionCommands = GetTableConstructionCommands(relationId);
+
+	tableConstructionCommands = list_concat(tableConstructionCommands,
+											GetViewCreationCommandsOfTable(relationId));
+
+	int spiResult = SPI_connect();
+	if (spiResult != SPI_OK_CONNECT)
+	{
+		ereport(ERROR, (errmsg("could not connect to SPI manager")));
+	}
+
+	if (PartitionedTable(relationId))
+	{
+		List *partitionList = PartitionList(relationId);
+		Oid partitionRelationId = InvalidOid;
+		foreach_oid(partitionRelationId, partitionList)
+		{
+			char *detachPartitionCommand = GenerateDetachPartitionCommand(
+				partitionRelationId);
+			char *attachPartitionCommand = GenerateAlterTableAttachPartitionCommand(
+				partitionRelationId);
+
+			spiResult = SPI_execute(detachPartitionCommand, false, 0);
+			if (spiResult != SPI_OK_UTILITY)
+			{
+				ereport(ERROR, (errmsg("could not run SPI query")));
+			}
+			tableCreationCommands = lappend(tableCreationCommands,
+											attachPartitionCommand);
+			UndistributeTable(partitionRelationId);
+		}
+	}
+
+	char *relationName = get_rel_name(relationId);
+	Oid schemaId = get_rel_namespace(relationId);
+	char *schemaName = get_namespace_name(schemaId);
+
+	char *tempName = pstrdup(relationName);
+	uint32 hashOfName = hash_any((unsigned char *) tempName, strlen(tempName));
+	AppendShardIdToName(&tempName, hashOfName);
+
+	char *tableCreationCommand = NULL;
+	foreach_ptr(tableCreationCommand, tableCreationCommands)
+	{
+		Node *parseTree = ParseTreeNode(tableCreationCommand);
+
+		/*
+		 * The GetTableCreationCommands also generates commands for creating
+		 * the extensions and servers necessary for creating the table. But
+		 * these are not needed here, since they already exist. So we filter
+		 * them out.
+		 */
+		if (IsA(parseTree, CreateStmt) ||
+			IsA(parseTree, AlterOwnerStmt) ||
+			IsA(parseTree, AlterTableStmt))
```

Comment:
It would be nicer to split GetTableCreationCommands  into two functions, one of which excludes the commands we don't care about here.